src/control/lib/spdk/spdk.go

```diff
@@ -67,18 +69,54 @@ func Rc2err(label string, rc C.int) error {
 	return fmt.Errorf("%s: %d", label, rc)
 }
 
+// backingAddress2VMD converts VMD backing device PCI addresses (with the VMD
+// address encoded in the domain component of the PCI address) back to the PCI
+// address of the VMD e.g. [5d0505:01:00.0, 5d0505:03:00.0] -> [0000:5d:05.5].
+//
+// Many assumptions are made as to the input and output PCI address structure in
+// the conversion.
+func backingAddress2VMD(log logging.Logger, pciAddrs *common.PCIAddressList) (*common.PCIAddressList, error) {
+	if pciAddrs == nil {
+		return nil, errors.New("revertBackingToVMD: nil pci address list")
+	}
+
+	outAddrs := common.PCIAddressList{}
+
+	for _, inAddr := range *pciAddrs {
+		if inAddr.Domain == "0000" {
+			outAddrs.Add(*inAddr)
+			continue
+		}
+
+		// assume non-zero pci address domain field indicates a vmd backing device with
+		// fixed length field
+		if len(inAddr.Domain) != vmdDomainLen {
+			return nil, errors.New("unexpected length of vmd domain")
+		}
+
+		vmdAddr := fmt.Sprintf("0000:%s:%s.%s", inAddr.Domain[:2], inAddr.Domain[2:4],
+			inAddr.Domain[5:])
+		if err := outAddrs.AddStrings(vmdAddr); err != nil {
+			log.Debugf("replacing backing device %s with vmd %s", inAddr, vmdAddr)
+			return nil, err
+		}
```

Comment:
I wonder if this code might be better placed in a method on the new PCIAddress type. It's already making some assumptions about VMD PCI format there.